python/lsst/qa/explorer/classifyObjects.py

```diff
@@ -0,0 +1,432 @@
+        classifier pickles and are used to ensure that the array of features is generated in the correct
+        order.
+
+        Apply both of the classifiers to the data
+
+        The columns added are:
+            PSF mangitudes in all the bands (in wavelength order), magnitudes, magPsf + band
+            PSF magntiude errors in all the bands, magnitudes, magPsfErr + band
+            CModel magnitudes in all the bands, magnitudes, magModel + band
+            CModel magnitude error in al the bands, magnitudes, magModelErr + band
+            Seeing in all the bands, arcsecs, seeing + band
+            Colors from all bands, magnitudes, first band - second band
+            Gradient of relationship between psf - cmodel and seeing, magnitudes/arcses, m
+            Intercept of relationship between psf - cmodel and seeing, magnitudes, b
+            Flux signal to noise in all bands, no units, fluxSN + band
+        These columns are added into a new index called `featuresSG`.
+        """
+        # Add required columns
+        cat = self.addSeeing(cat, filters)
+        cat = self.addColors(cat, filters)
+        cat = self.addMB(cat, filters)
+        cat = self.addSNFlux(cat, filters)
+
+        featuresMorph = np.zeros((len(colsToUseMorph), len(cat)))
+        for (i, col) in enumerate(colsToUseMorph):
+            featuresMorph[i] = cat["featuresSGSep"][col].data
+
+        # Need to take out the nans but remember which rows they are and assign nans to the probability and
+        # add a flag
+        featureSumsMorph = np.sum(featuresMorph, axis=0)
+        cleanIdsMorph = np.where((np.isfinite(featureSumsMorph)))[0]
+
+        cleanFeaturesMorph = featuresMorph[:, cleanIdsMorph]
+
+        flagMorph = np.ones(len(cat))
+        flagMorph[cleanIdsMorph] = 0
+
+        probsOutMorph = np.array([np.nan]*len(cat))
+
+        probsMorph = clfMorph.predict_proba(cleanFeaturesMorph.T)
+        probsOutMorph[cleanIdsMorph] = probsMorph[:, 0]
+
+        features = np.zeros((len(colsToUse), len(cat)))
+
+        for (i, col) in enumerate(colsToUse):
+            features[i] = cat["featuresSGSep"][col].data
+
+        # Need to take out the nans but remember which rows they are and assign nans to the probability and
+        # add a flag
+        featureSums = np.sum(features, axis=0)
+        cleanIds = np.where((np.isfinite(featureSums)))[0]
+
+        cleanFeatures = features[:, cleanIds]
+
+        flag = np.ones(len(cat))
+        flag[cleanIds] = 0
+
+        probsOut = np.array([np.nan]*len(cat))
+
+        probs = clf.predict_proba(cleanFeatures.T)
+        probsOut[cleanIds] = probs[:, 0]
+
+        for band in filters:
+            cat[band, "starProbColors"] = probsOut
+            cat[band, "starProbMorph"] = probsOutMorph
+            cat[band, "starProbColorsFlag"] = flag
+            cat[band, "starProbMorphFlag"] = flagMorph
```

Comment:
At the moment the table is setup in such a way that you can read in only each band individually if you want to. Therefore a number of columns, such as tractId, are duplicated into the table for each band then people who only want access to the z band can still have the band independent information.

The colour based probability uses all the colours and is thus the same for all the bands. It is duplicated for the same reasons as above.

I do not know if there is a better way of doing this and it might be that in the future there is a table of band independent quantities or one non-subdivided file that includes a sub set of all the columns for all the bands and then it would not need to be duplicated.